commands/voiceChannelManager.js

```diff
@@ -0,0 +1,251 @@
+let ftl = require('../lib/ftl');
+
+// Cleanup intervals under 5 minutes will not be reliable.
+const CLEANUP_INTERVAL_MINUTES = 5;
+const CHANNEL_TYPE = { TEXT: 0, VOICE: 2, GROUP: 4 };
+const EMOJIS = {
+  CHANNEL_PREFIX: '⚡',
+  SUCCESS: '✅',
+  CONFIRM_DELETE: '🔥',
+  CANCEL_DELETE: '🙊',
+  CLEANUP: '✂',
+};
+
+// string length is number of bytes, and many emoji are not one byte. Regex can handle that better.
+const oneChar = /^.$/u;
+for (const [name, emoji] of Object.entries(EMOJIS)) {
+  if (!emoji.match(oneChar)) {
+    throw new Error(`Emojis must be just one character, ${name} is too long`);
+  }
+}
+
+module.exports.register = ({ bot }) => {
+  new VcCommand({ bot });
+};
+
+class VcCommand {
+  constructor({ bot }) {
+    this.bot = bot;
+    this.trackedChannels = new Map();
+    this._targetGroup = null;
+
+    const vcCommand = bot.registerCommand('voice-channel', ftl('error-unknown-subcommand'), {
+      aliases: ['vc'],
+      argsRequired: true,
+      guildOnly: true,
+      ftlVariables: {
+        channelTimeoutMinutes: CLEANUP_INTERVAL_MINUTES * 2,
+        channelCleanupEmoji: EMOJIS.CLEANUP,
+      },
+    });
+
+    vcCommand.registerSubcommand('create', (...args) => this.subcommandCreate(...args), {
+      argsRequired: true,
+      guildOnly: true,
+    });
+
+    vcCommand.registerSubcommand('delete', (...args) => this.subcommandDelete(...args), {
+      argsRequired: true,
+      guildOnly: true,
+    });
+
+    vcCommand.registerSubcommand(
+      'delete-all',
+      (...args) => this.subcommandDeleteAllWarning(...args),
+      {
+        guildOnly: true,
+        reactionButtons: [
+          {
+            emoji: EMOJIS.CONFIRM_DELETE,
+            type: 'edit',
+            response: (...args) => this.reactionCommandDeleteAll(...args),
+          },
+          {
+            emoji: EMOJIS.CANCEL_DELETE,
+            type: 'cancel',
+            response: ftl('voice-channel-cmd-delete-all-canceled'),
+          },
+        ],
+      },
+    );
+
+    setInterval(() => this.cleanupTask(), CLEANUP_INTERVAL_MINUTES * 60 * 1000);
+    // Check for channel updates right away, with a short delay to allow the bot to log in.
+    setTimeout(() => this.cleanupTask(), 1000);
+  }
+
+  async getChannelGroup(guild) {
+    if (!this._targetGroup) {
+      const channels = await this.bot.getRESTGuildChannels(guild.id);
+      let targetGroup = channels.find((channel) => channel.name.startsWith(EMOJIS.CHANNEL_PREFIX));
+      if (!targetGroup) {
+        targetGroup = await this.bot.createChannel(
+          guild.id,
+          `${EMOJIS.CHANNEL_PREFIX} AUTO CHANNELS`,
+          CHANNEL_TYPE.GROUP,
+          {
+            reason: `To hold !vc command channesl`,
+          },
+        );
+      }
+      this._targetGroup = targetGroup;
+    }
+    return this._targetGroup;
+  }
+
+  async findChannel(guild, needle) {
```

Comment:
needle? as in needle in a hay stack?

Reply:
Yep, that's exactly the metaphor. I spent a while working on some search-engine-like code, and picked it up there.